File: js/src/sandbox_sidecar.ts
import { ClientError, Status } from "nice-grpc";

import {
  GenericResult,
  GenericResult_GenericStatus,
} from "../proto/modal_proto/api";
import {
  TaskContainerCreateRequest,
  TaskContainerGetRequest,
  TaskContainerInfo,
  TaskContainerListRequest,
  TaskContainerTerminateRequest,
  TaskContainerWaitRequest,
} from "../proto/modal_proto/task_command_router";
import {
  ContainerProcess,
  getReturnCode,
  validateExecArgs,
  validateWorkdir,
  type SandboxExecParams,
} from "./sandbox";
import { SandboxFilesystem } from "./sandbox_fs";
import type { TaskCommandRouterClientImpl } from "./task_command_router_client";
import {
  AlreadyExistsError,
  InternalFailure,
  InvalidError,
  NotFoundError,
} from "./errors";
import type { Image } from "./image";
import type { ModalClient } from "./client";
import {
  collectSecretIds,
  hydrateSecrets,
  splitEnvDictAndResolvableSecrets,
  validateEnvVarKeys,
  type Secret,
} from "./secret";

/** Reserved name of a Sandbox's main container. */
const MAIN_CONTAINER_NAME = "main";

/**
 * Per-RPC server-side wait timeout (in seconds) used while polling for a
 * sidecar container's terminal status.
 */
const CONTAINER_WAIT_POLL_TIMEOUT_SECONDS = 10;

type SandboxSidecarCommandRouter = Pick<
  TaskCommandRouterClientImpl,
  | "containerCreate"
  | "containerGet"
  | "containerList"
  | "containerTerminate"
  | "containerWait"
>;

type SandboxSidecarAccess = {
  client: ModalClient;
  exec(
    command: string[],
    params: SandboxExecParams | undefined,
    containerId: string,
  ): Promise<ContainerProcess>;
  commandRouter(): Promise<[string, SandboxSidecarCommandRouter]>;
};

/** Options for {@link SidecarService#create SidecarService.create()}. */
export type SidecarCreateParams = {
  /** Command to run in the sidecar container on startup. */
  command?: string[];
  /** Environment variables to set in the sidecar container. */
  env?: Record<string, string>;
  /** {@link Secret}s to inject into the sidecar container as environment variables. */
  secrets?: Secret[];
  /** Working directory of the sidecar container. */
  workdir?: string;
};

/** Options for {@link SidecarService#get SidecarService.get()}. */
export type SidecarGetParams = {
  /** If true, return the latest container with the name even if it has terminated. */
  includeTerminated?: boolean;
};

/** Options for {@link SidecarService#list SidecarService.list()}. */
export type SidecarListParams = {
  /** If true, include terminated containers. */
  includeTerminated?: boolean;
};

/** Options for {@link SidecarContainer#exec SidecarContainer.exec()}. */
export type SidecarExecParams = SandboxExecParams;

/** Options for {@link SidecarContainer#terminate SidecarContainer.terminate()}. */
export type SidecarTerminateParams = {
  /** If true, wait for the sidecar container to terminate. */
  wait?: boolean;
};

function validateSidecarName(name: string): void {
  if (name === "") {
    throw new InvalidError("sidecar name must not be empty");
  }
  if (name === MAIN_CONTAINER_NAME) {
    throw new InvalidError(
      `the name "${MAIN_CONTAINER_NAME}" is reserved for the Sandbox's main ` +
        `container. Use the Sandbox methods directly to interact with it`,
    );
  }
}

function sidecarContainerFromProto(
  access: SandboxSidecarAccess,
  info: TaskContainerInfo,
): SidecarContainer {
  return new SidecarContainer(
    access,
    info.containerId,
    info.containerName,
    info.result,
  );
}

/**
 * Creates and manages sidecar containers inside a Sandbox.
 *
 * EXPERIMENTAL: the API is subject to change.
 */
export class SidecarService {
  readonly #access: SandboxSidecarAccess;

  /** @internal @hidden */
  constructor(access: SandboxSidecarAccess) {
    this.#access = access;
  }

  /**
   * Start a new sidecar container in the Sandbox. The {@link Image} must
   * already be built by calling {@link Image#build Image.build()} before it is
   * passed to `create`.
   */
  async create(
    name: string,
    image: Image,
    params?: SidecarCreateParams,
  ): Promise<SidecarContainer> {
    validateSidecarName(name);
    if (!image || image.imageId === "") {
      throw new InvalidError(
        "sidecar image must already be built. Call image.build(app) first " +
          "or use client.images.fromId(...)",
      );
    }
    const command = params?.command ?? [];
    validateExecArgs(command);
    validateWorkdir(params?.workdir);

    // Locally-created Secrets (fromObject) are passed directly to the worker as
    // environment variables, so only the remaining Secrets need hydrating. This
    // avoids a SecretGetOrCreate round-trip for env-dict Secrets.
    validateEnvVarKeys(params?.env ?? {});
    const [envDict, resolvableSecrets] = splitEnvDictAndResolvableSecrets(
      params?.secrets ?? [],
    );
    Object.assign(envDict, params?.env ?? {});
    await hydrateSecrets(this.#access.client, resolvableSecrets);
    const secretIds = collectSecretIds(resolvableSecrets);

    const [taskId, client] = await this.#access.commandRouter();

    let resp;
    try {
      resp = await client.containerCreate(
        TaskContainerCreateRequest.create({
          taskId,
          containerName: name,
          imageId: image.imageId,
          args: command,
          env: envDict,
          workdir: params?.workdir ?? "",
          secretIds,
        }),
      );
    } catch (err) {
      if (err instanceof ClientError) {
        if (err.code === Status.ALREADY_EXISTS) {
          throw new AlreadyExistsError(err.details || err.message);
        }
        if (err.code === Status.INVALID_ARGUMENT) {
          throw new InvalidError(err.details || err.message);
        }
      }
      throw err;
    }

    return new SidecarContainer(
      this.#access,
      resp.containerId,
      resp.containerName || name,
    );
  }

  /** Return a sidecar container by name. */
  async get(
    name: string,
    params?: SidecarGetParams,
  ): Promise<SidecarContainer> {
    validateSidecarName(name);

    const [taskId, client] = await this.#access.commandRouter();

    let resp;
    try {
      resp = await client.containerGet(
        TaskContainerGetRequest.create({
          taskId,
          containerName: name,
          includeTerminated: params?.includeTerminated ?? false,
        }),
      );
    } catch (err) {
      if (err instanceof ClientError && err.code === Status.NOT_FOUND) {
        throw new NotFoundError(`Sidecar container "${name}" not found`);
      }
      throw err;
    }
    if (!resp.container) {
      throw new InternalFailure(
        `server returned no container for sidecar "${name}"`,
      );
    }
    return sidecarContainerFromProto(this.#access, resp.container);
  }

  /** Return all sidecar containers (not including the main container). */
  async list(params?: SidecarListParams): Promise<SidecarContainer[]> {
    const [taskId, client] = await this.#access.commandRouter();

    const resp = await client.containerList(
      TaskContainerListRequest.create({
        taskId,
        includeTerminated: params?.includeTerminated ?? false,
      }),
    );

    return resp.containers
      .filter((info) => info.containerName !== MAIN_CONTAINER_NAME)
      .map((info) => sidecarContainerFromProto(this.#access, info));
  }
}

/**
 * A handle to a sidecar container running in a Sandbox.
 *
 * EXPERIMENTAL: the API is subject to change.
 */
export class SidecarContainer {
  /** The fully qualified container ID. */
  readonly containerId: string;
  /** The logical name of the container within the Sandbox. */
  readonly containerName: string;

  readonly #access: SandboxSidecarAccess;
  #result?: GenericResult;
  #filesystem?: SandboxFilesystem;

  /** @internal @hidden */
  constructor(
    access: SandboxSidecarAccess,
    containerId: string,
    containerName: string,
    result?: GenericResult,
  ) {
    this.#access = access;
    this.containerId = containerId;
    this.containerName = containerName;
    this.#result = result;
  }

  async exec(
    command: string[],
    params?: SidecarExecParams & { mode?: "text" },
  ): Promise<ContainerProcess<string>>;

  async exec(
    command: string[],
    params: SidecarExecParams & { mode: "binary" },
  ): Promise<ContainerProcess<Uint8Array>>;

  /** Run a command in the sidecar container and return the process handle. */
  async exec(
    command: string[],
    params?: SidecarExecParams,
  ): Promise<ContainerProcess> {
    return this.#access.exec(command, params, this.containerId);
  }

  /** Namespace for filesystem APIs scoped to this sidecar container. */
  get filesystem(): SandboxFilesystem {
    if (!this.#filesystem) {
      this.#filesystem = new SandboxFilesystem((command, params) =>
        this.#access.exec(command, params, this.containerId),
      );
    }
    return this.#filesystem;
  }

  /** Block until the sidecar container exits, and return its exit code. */
  async wait(): Promise<number> {
    if (
      this.#result &&
      this.#result.status !==
        GenericResult_GenericStatus.GENERIC_STATUS_UNSPECIFIED
    ) {
      return getReturnCode(this.#result) ?? 0;
    }

    const [taskId, client] = await this.#access.commandRouter();
    while (true) {
      const resp = await client.containerWait(
        TaskContainerWaitRequest.create({
          taskId,
          containerId: this.containerId,
          timeout: CONTAINER_WAIT_POLL_TIMEOUT_SECONDS,
        }),
      );
      const result = resp.result;
      if (
        !result ||
        result.status === GenericResult_GenericStatus.GENERIC_STATUS_UNSPECIFIED
      ) {
        continue;
      }
      this.#result = result;
      return getReturnCode(result) ?? 0;
    }
  }

  /**
   * Check if the sidecar container has finished running.
   *
   * Returns `null` if the container is still running, else the exit code.
   */
  async poll(): Promise<number | null> {
    if (
      this.#result &&
      this.#result.status !==
        GenericResult_GenericStatus.GENERIC_STATUS_UNSPECIFIED
    ) {
      return getReturnCode(this.#result);
    }

    const [taskId, client] = await this.#access.commandRouter();
    const resp = await client.containerWait(
      TaskContainerWaitRequest.create({
        taskId,
        containerId: this.containerId,
        timeout: 0,
      }),
    );
    const result = resp.result;
    if (
      result &&
      result.status !== GenericResult_GenericStatus.GENERIC_STATUS_UNSPECIFIED
    ) {
      this.#result = result;
    }
    return getReturnCode(result);
  }

  async terminate(): Promise<void>;
  async terminate(params: { wait: true }): Promise<number>;
  /**
   * Stop the sidecar container.
   *
   * The returned exit code is only meaningful when `wait` is true.
   */
  async terminate(params?: SidecarTerminateParams): Promise<number | void> {
    const [taskId, client] = await this.#access.commandRouter();
    await client.containerTerminate(
      TaskContainerTerminateRequest.create({
        taskId,
        containerId: this.containerId,
      }),
    );
    if (params?.wait) {
      return this.wait();
    }
  }
}

File: go/sandbox_sidecar.go
package modal

import (
	"context"
	"fmt"
	"sync"
	"time"

	pb "github.com/modal-labs/modal-client/go/proto/modal_proto"
	"google.golang.org/grpc/codes"
	"google.golang.org/grpc/status"
)

// mainContainerName is the reserved name of a Sandbox's main container.
const mainContainerName = "main"

// containerWaitPollTimeout is the per-RPC server-side wait timeout used while
// polling for a sidecar's terminal status.
const containerWaitPollTimeout = 10 * time.Second

// SidecarService creates and manages sidecar containers inside a Sandbox.
//
// EXPERIMENTAL: the API is subject to change.
type SidecarService interface {
	// Create starts a new sidecar container in the Sandbox. The Image must
	// already be built by calling [Image.Build] before it's passed to Create.
	Create(ctx context.Context, name string, image *Image, params *SidecarCreateParams) (*SidecarContainer, error)
	// Get returns a sidecar container by name.
	Get(ctx context.Context, name string, params *SidecarGetParams) (*SidecarContainer, error)
	// List returns all sidecar containers (not including the main container).
	List(ctx context.Context, params *SidecarListParams) ([]*SidecarContainer, error)
}

type sidecarServiceImpl struct{ sandbox *Sandbox }

// SidecarCreateParams holds options for creating a sidecar container.
type SidecarCreateParams struct {
	// Command to run in the sidecar container on startup.
	Command []string
	// Env are environment variables to set in the sidecar container.
	Env map[string]string
	// Secrets to inject into the sidecar container as environment variables.
	Secrets []*Secret
	// Workdir sets the working directory of the sidecar container.
	Workdir string
}

// SidecarGetParams holds options for retrieving a sidecar by name.
type SidecarGetParams struct {
	IncludeTerminated bool
}

// SidecarListParams holds options for listing sidecars.
type SidecarListParams struct {
	IncludeTerminated bool
}

// SidecarExecParams holds options for [SidecarContainer.Exec].
type SidecarExecParams SandboxExecParams

// SidecarWaitParams holds options for [SidecarContainer.Wait].
type SidecarWaitParams struct{}

// SidecarPollParams holds options for [SidecarContainer.Poll].
type SidecarPollParams struct{}

// SidecarTerminateParams holds options for [SidecarContainer.Terminate].
type SidecarTerminateParams struct {
	// Wait, when true, will wait for the sidecar container to terminate.
	Wait bool
}

func validateSidecarName(name string) error {
	if name == "" {
		return InvalidError{Exception: "sidecar name must not be empty"}
	}
	if name == mainContainerName {
		return InvalidError{Exception: fmt.Sprintf("the name %q is reserved for the Sandbox's main container. Use the Sandbox methods directly to interact with it", mainContainerName)}
	}
	return nil
}

func (s *sidecarServiceImpl) Create(ctx context.Context, name string, image *Image, params *SidecarCreateParams) (*SidecarContainer, error) {
	if err := validateSidecarName(name); err != nil {
		return nil, err
	}
	if image == nil || image.ImageID == "" {
		return nil, InvalidError{Exception: "sidecar image must already be built. Call image.Build(ctx, app) first or use Images.FromID(...)"}
	}
	if params == nil {
		params = &SidecarCreateParams{}
	}
	if err := validateWorkdir(params.Workdir); err != nil {
		return nil, err
	}
	if err := validateExecArgs(params.Command); err != nil {
		return nil, err
	}

	// Locally-created Secrets (FromMap) are passed directly to the worker as
	// environment variables, so only the remaining Secrets need hydrating. This
	// avoids a SecretGetOrCreate round-trip for env-dict Secrets.
	envDict, resolvableSecrets := splitEnvDictAndResolvableSecrets(params.Secrets)
	for k, v := range params.Env {
		if err := validateEnvVarName(k); err != nil {
			return nil, err
		}
		envDict[k] = v
	}
	if err := hydrateSecrets(ctx, s.sandbox.client, resolvableSecrets); err != nil {
		return nil, err
	}
	secretIds, err := collectSecretIDs(resolvableSecrets)
	if err != nil {
		return nil, err
	}

	taskID, client, err := s.sandbox.getCommandRouter(ctx)
	if err != nil {
		return nil, err
	}

	req := pb.TaskContainerCreateRequest_builder{
		TaskId:        taskID,
		ContainerName: name,
		ImageId:       image.ImageID,
		Args:          params.Command,
		Env:           envDict,
		Workdir:       params.Workdir,
		SecretIds:     secretIds,
	}.Build()

	resp, err := client.ContainerCreate(ctx, req)
	if err != nil {
		if st, ok := status.FromError(err); ok {
			switch st.Code() {
			case codes.AlreadyExists:
				return nil, AlreadyExistsError{Exception: st.Message()}
			case codes.InvalidArgument:
				return nil, InvalidError{Exception: st.Message()}
			}
		}
		return nil, err
	}

	containerName := resp.GetContainerName()
	if containerName == "" {
		containerName = name
	}
	s.sandbox.client.logger.DebugContext(ctx, "Created SidecarContainer",
		"container_id", resp.GetContainerId(),
		"container_name", containerName,
		"sandbox_id", s.sandbox.SandboxID)
	return newSidecarContainer(s.sandbox, resp.GetContainerId(), containerName, nil), nil
}

func (s *sidecarServiceImpl) Get(ctx context.Context, name string, params *SidecarGetParams) (*SidecarContainer, error) {
	if err := validateSidecarName(name); err != nil {
		return nil, err
	}
	if params == nil {
		params = &SidecarGetParams{}
	}

	taskID, client, err := s.sandbox.getCommandRouter(ctx)
	if err != nil {
		return nil, err
	}

	resp, err := client.ContainerGet(ctx, pb.TaskContainerGetRequest_builder{
		TaskId:            taskID,
		ContainerName:     name,
		IncludeTerminated: params.IncludeTerminated,
	}.Build())
	if err != nil {
		if st, ok := status.FromError(err); ok && st.Code() == codes.NotFound {
			return nil, NotFoundError{Exception: fmt.Sprintf("Sidecar container %q not found", name)}
		}
		return nil, err
	}
	container := resp.GetContainer()
	if container == nil {
		return nil, ExecutionError{Exception: fmt.Sprintf("server returned no container for sidecar %q", name)}
	}
	return sidecarContainerFromProto(s.sandbox, container), nil
}

func (s *sidecarServiceImpl) List(ctx context.Context, params *SidecarListParams) ([]*SidecarContainer, error) {
	if params == nil {
		params = &SidecarListParams{}
	}

	taskID, client, err := s.sandbox.getCommandRouter(ctx)
	if err != nil {
		return nil, err
	}

	resp, err := client.ContainerList(ctx, pb.TaskContainerListRequest_builder{
		TaskId:            taskID,
		IncludeTerminated: params.IncludeTerminated,
	}.Build())
	if err != nil {
		return nil, err
	}

	containers := resp.GetContainers()
	out := make([]*SidecarContainer, 0, len(containers))
	for _, info := range containers {
		if info.GetContainerName() == mainContainerName {
			continue
		}
		out = append(out, sidecarContainerFromProto(s.sandbox, info))
	}
	return out, nil
}

// SidecarContainer is a handle to a sidecar container running in a Sandbox.
type SidecarContainer struct {
	// ContainerID is the fully qualified container ID.
	ContainerID string
	// ContainerName is the logical name of the container within the Sandbox.
	ContainerName string

	// Filesystem provides high-level filesystem operations for this container.
	Filesystem *SandboxFilesystem

	sandbox *Sandbox

	resultMu sync.Mutex
	result   *pb.GenericResult
}

func newSidecarContainer(sandbox *Sandbox, id, name string, result *pb.GenericResult) *SidecarContainer {
	c := &SidecarContainer{
		ContainerID:   id,
		ContainerName: name,
		sandbox:       sandbox,
		result:        result,
	}
	c.Filesystem = &SandboxFilesystem{sandbox: c, logger: sandbox.client.logger}
	return c
}

func sidecarContainerFromProto(sandbox *Sandbox, info *pb.TaskContainerInfo) *SidecarContainer {
	if info == nil {
		return nil
	}
	var result *pb.GenericResult
	if info.HasResult() {
		result = info.GetResult()
	}
	return newSidecarContainer(sandbox, info.GetContainerId(), info.GetContainerName(), result)
}

// Exec runs a command in the sidecar container and returns the process handle.
func (c *SidecarContainer) Exec(ctx context.Context, command []string, params *SidecarExecParams) (*ContainerProcess, error) {
	return c.sandbox.execInternal(ctx, command, (*SandboxExecParams)(params), c.ContainerID)
}

// execForFilesystem satisfies the sandboxForFilesystem interface for SidecarContainer.
func (c *SidecarContainer) execForFilesystem(ctx context.Context, command []string, params *SandboxExecParams) (*ContainerProcess, error) {
	return c.sandbox.execInternal(ctx, command, params, c.ContainerID)
}

// Wait blocks until the sidecar container exits, and returns its exit code.
func (c *SidecarContainer) Wait(ctx context.Context, _ *SidecarWaitParams) (int, error) {
	c.resultMu.Lock()
	cached := c.result
	c.resultMu.Unlock()
	if code := getReturnCode(cached); code != nil {
		return *code, nil
	}

	taskID, client, err := c.sandbox.getCommandRouter(ctx)
	if err != nil {
		return 0, err
	}

	for {
		if err := ctx.Err(); err != nil {
			return 0, err
		}
		resp, err := client.ContainerWait(ctx, pb.TaskContainerWaitRequest_builder{
			TaskId:      taskID,
			ContainerId: c.ContainerID,
			Timeout:     float32(containerWaitPollTimeout.Seconds()),
		}.Build())
		if err != nil {
			return 0, err
		}
		result := resp.GetResult()
		if result == nil || result.GetStatus() == pb.GenericResult_GENERIC_STATUS_UNSPECIFIED {
			continue
		}
		c.resultMu.Lock()
		c.result = result
		c.resultMu.Unlock()
		if code := getReturnCode(result); code != nil {
			return *code, nil
		}
		return 0, nil
	}
}

// Poll checks if the sidecar container has finished running.
// Returns nil if the container is still running, else returns the exit code.
func (c *SidecarContainer) Poll(ctx context.Context, _ *SidecarPollParams) (*int, error) {
	c.resultMu.Lock()
	cached := c.result
	c.resultMu.Unlock()
	if code := getReturnCode(cached); code != nil {
		return code, nil
	}

	taskID, client, err := c.sandbox.getCommandRouter(ctx)
	if err != nil {
		return nil, err
	}

	resp, err := client.ContainerWait(ctx, pb.TaskContainerWaitRequest_builder{
		TaskId:      taskID,
		ContainerId: c.ContainerID,
		Timeout:     0,
	}.Build())
	if err != nil {
		return nil, err
	}
	result := resp.GetResult()
	if result == nil {
		return nil, nil
	}
	if result.GetStatus() != pb.GenericResult_GENERIC_STATUS_UNSPECIFIED {
		c.resultMu.Lock()
		c.result = result
		c.resultMu.Unlock()
	}
	return getReturnCode(result), nil
}

// Terminate stops the sidecar container.
//
// The returned exit code is only meaningful when Wait is true.
func (c *SidecarContainer) Terminate(ctx context.Context, params *SidecarTerminateParams) (int, error) {
	if params == nil {
		params = &SidecarTerminateParams{}
	}

	taskID, client, err := c.sandbox.getCommandRouter(ctx)
	if err != nil {
		return 0, err
	}

	if err := client.ContainerTerminate(ctx, pb.TaskContainerTerminateRequest_builder{
		TaskId:      taskID,
		ContainerId: c.ContainerID,
	}.Build()); err != nil {
		return 0, err
	}

	if !params.Wait {
		return 0, nil
	}
	return c.Wait(ctx, nil)
}
